File: dotnet/system/workspace/csharp/allors.workspace.domain/istrategy.cs
// <copyright file="Object.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System.Collections.Generic;
    using Meta;

    public interface IStrategy
    {
        ISession Session { get; }

        IObject Object { get; }

        IClass Class { get; }

        long Id { get; }

        long Version { get; }

        void Reset();

        IReadOnlyList<IDiff> Diff();

        bool CanRead(IRoleType roleType);

        bool CanWrite(IRoleType roleType);

        bool CanExecute(IMethodType methodType);

        bool ExistRole(IRoleType roleType);

        object GetRole(IRoleType roleType);

        void SetRole(IRoleType roleType, object value);

        void RemoveRole(IRoleType roleType);

        object GetUnitRole(IRoleType roleType);

        void SetUnitRole(IRoleType roleType, object value);

        T GetCompositeRole<T>(IRoleType roleType) where T : class, IObject;

        void SetCompositeRole<T>(IRoleType roleType, T value) where T : class, IObject;

        IEnumerable<T> GetCompositesRole<T>(IRoleType roleType) where T : class, IObject;

        void AddCompositesRole<T>(IRoleType roleType, T value) where T : class, IObject;

        void RemoveCompositesRole<T>(IRoleType roleType, T value) where T : class, IObject;

        void SetCompositesRole<T>(IRoleType roleType, in IEnumerable<T> role) where T : class, IObject;

        T GetCompositeAssociation<T>(IAssociationType associationType) where T : class, IObject;

        IEnumerable<T> GetCompositesAssociation<T>(IAssociationType associationType) where T : class, IObject;
    }
}

File: dotnet/system/workspace/typescript/libs/workspace/domain/system/src/lib/istrategy.ts
import { Class, RelationType, AssociationType,  RoleType, MethodType} from "@allors/workspace/meta/system";
import { IObject } from './IObject';
import { ISession } from './ISession';
import { UnitTypes } from './Types';

export interface IStrategy {
  object: IObject;

  class: Class;

  id: number;

  session: ISession;

  diff(): RelationType[];

  canRead(roleType: RoleType): boolean;

  canWrite(roleType: RoleType): boolean;

  canExecute(methodType: MethodType): boolean;

  exist(roleType: RoleType): boolean;

  get(roleType: RoleType): unknown;

  getUnit(roleType: RoleType): UnitTypes;

  getComposite<T>(roleType: RoleType): T;

  getComposites<T>(roleType: RoleType): T[];

  set(roleType: RoleType, value: unknown): void;

  setUnit(roleType: RoleType, value: UnitTypes): void;

  setComposite<T>(roleType: RoleType, value: T): void;

  setComposites<T>(roleType: RoleType, value: T[]): void;

  add<T>(roleType: RoleType, value: T): void;

  remove<T>(roleType: RoleType, value: T): void;

  removeAll(roleType: RoleType): void;

  getCompositeAssociation<T>(associationType: AssociationType): T;

  getCompositesAssociation<T>(associationType: AssociationType): T[];

  reset(): void;
}
